go/worker/sentry/grpc/worker.go

```diff
@@ -0,0 +1,199 @@
+package grpc
+
+import (
+	"context"
+	"sync"
+
+	"github.com/oasislabs/oasis-core/go/common/accessctl"
+	cmnGrpc "github.com/oasislabs/oasis-core/go/common/grpc"
+	"github.com/oasislabs/oasis-core/go/common/grpc/auth"
+	"github.com/oasislabs/oasis-core/go/common/grpc/policy"
+	policyAPI "github.com/oasislabs/oasis-core/go/common/grpc/policy/api"
+	"github.com/oasislabs/oasis-core/go/common/identity"
+	"github.com/oasislabs/oasis-core/go/common/logging"
+	"github.com/oasislabs/oasis-core/go/common/service"
+	storageAPI "github.com/oasislabs/oasis-core/go/storage/api"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var _ service.BackgroundService = (*Worker)(nil)
+
+// Worker is a gRPC sentry node worker proxiying gRPC requets to upstream node.
+type Worker struct {
+	sync.RWMutex
+
+	enabled bool
+
+	ctx       context.Context
+	cancelCtx context.CancelFunc
+
+	initCh chan struct{}
+	stopCh chan struct{}
+	quitCh chan struct{}
+
+	logger *logging.Logger
+
+	policyWatcher policyAPI.PolicyWatcherClient
+	// Per service polcy checkers.
+	grpcPolicyCheckers map[cmnGrpc.ServiceName]*policy.DynamicRuntimePolicyChecker
+
+	*upstreamConn
+
+	grpc     *cmnGrpc.Server
+	identity *identity.Identity
+}
+
+type upstreamConn struct {
+	conn              *grpc.ClientConn
+	resolverCleanupCb func()
+}
+
+// TODO: this whole authFunction is currently hardcoded to only supports 'storage' backend,
+//       so that i was able to do test. Sentry GRPC worker should be upstream agnostic,
```

Comment:
Yeah that would be ideal. What exactly needs to change here? So the most problematic thing seems to be extracting "namespace" from request? How could this be handled -- each service method would need to define how to extract the namespace.

* Either the upstream provides a "key path" in the CBOR document for each method (so the proxy would unmarshal into an `interface{}` and extract the path, then parse it as a namespace), or
* the proxy should be expected to know all services (registry of all services by name?) and have "namespace extractor" methods registered.

Maybe the second one is simpler as we can assume that the proxy also knows all services (as it is just another oasis-node). The first one would allow for simpler proxies though (they would only need to know how to decode and traverse CBOR).

Reply:
done something in this direction:
> the proxy should be expected to know all services (registry of all services by name?) and have "namespace extractor" methods registered.

by creating a registry of method types and namespace extractors. please take another look

Reply:
Cool, yeah this looks like a nice approach.